app/iot/service.py

```diff
@@ -1,49 +1,75 @@
 import random
 import string
-from typing import Protocol
+import asyncio
+from typing import Protocol, Awaitable, Any
 
-from .message import Message, MessageType
+from app.iot.message import Message, MessageType
 
 
 def generate_id(length: int = 8) -> str:
     return "".join(random.choices(string.ascii_uppercase, k=length))
 
 
 # Protocol is very similar to ABC, but uses duck typing
-# so devices should not inherit for it (if it walks like a duck, and quacks like a duck, it's a duck)
+# so devices should not inherit for it
+# (if it walks like a duck, and quacks like a duck, it's a duck)
 class Device(Protocol):
-    def connect(self) -> None:
-        ...  # Ellipsis - similar to "pass", but sometimes has different meaning
+    async def connect(self) -> None:
+        ...
+        # Ellipsis - similar to "pass", but sometimes has different meaning
 
-    def disconnect(self) -> None:
+    async def disconnect(self) -> None:
         ...
 
-    def send_message(self, message_type: MessageType, data: str) -> None:
+    async def send_message(self, message_type: MessageType, data: str) -> None:
         ...
 
 
 class IOTService:
     def __init__(self) -> None:
         self.devices: dict[str, Device] = {}
 
-    def register_device(self, device: Device) -> str:
-        device.connect()
+    async def register_device(self, device: Device) -> str:
+        await device.connect()
         device_id = generate_id()
         self.devices[device_id] = device
         return device_id
 
-    def unregister_device(self, device_id: str) -> None:
-        self.devices[device_id].disconnect()
+    async def unregister_device(self, device_id: str) -> None:
+        await self.devices[device_id].disconnect()
         del self.devices[device_id]
 
     def get_device(self, device_id: str) -> Device:
         return self.devices[device_id]
 
-    def run_program(self, program: list[Message]) -> None:
+    async def run_sequence(self, *functions: Awaitable[Any]) -> None:
+        for function in functions:
+            await function
+
+    async def run_parallel(self, *functions: Awaitable[Any]) -> None:
+        await asyncio.gather(*functions)
+
+    async def run_program(self, program: list[Message]) -> None:
         print("=====RUNNING PROGRAM======")
-        for msg in program:
-            self.send_msg(msg)
+        flush_clean = [
+            msg for msg in program
+            if msg.msg_type in {MessageType.FLUSH, MessageType.CLEAN}
+        ]
+        other_commands = [
+            msg for msg in program
+            if msg.msg_type not in {MessageType.FLUSH, MessageType.CLEAN}
+        ]
+
+        if flush_clean:
+            await asyncio.gather(
+                *(self.send_msg(msg) for msg in other_commands),
+                self.run_sequence(*(self.send_msg(msg) for msg in flush_clean))
+            )
+        else:
+            await asyncio.gather(
+                *(self.send_msg(msg) for msg in other_commands)
+            )
```

Comment:
actually not really understood why this code exists, no matter what operation, because we are simply running the program, so smth like code lower - enough for your functionality
```
await asyncio.gather(*[self.send_msg(msg) for msg in program])
```